File: subsets/subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(result, nums, new ArrayList<Integer>(), 0);
        return result;
    }
    
    void backtrack(List<List<Integer>> result, int[] nums, List<Integer> temp, int start) {
        result.add(new ArrayList<>(temp));
        for (int i = start; i < nums.length; i++) {
            temp.add(nums[i]);
            backtrack(result, nums, temp, i+1);
            temp.remove(temp.size()-1);
        }
    }
}

File: subsets/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = []
        self.backtrack(result, nums, 0, [])
        return result

    def backtrack(self, result, nums, start, temp):
        result.append(copy.copy(temp))
        for index in range(start, len(nums)):
            temp.append(nums[index])
            self.backtrack(result, nums, index + 1, temp)
            del temp[-1]